promql/engine.go

```diff
@@ -1088,6 +1088,146 @@ func (ev *evaluator) rangeEval(funcCall func([]parser.Value, *EvalNodeHelper) (V
+		if e != nil && e.Type() != parser.ValueTypeString {
+			// ev.currentSamples will be updated to the correct value within the ev.eval call.
+			val, ws := ev.eval(e)
+			warnings = append(warnings, ws...)
+			matrixes[i] = val.(Matrix)
+
+			// Keep a copy of the original point slices so that they
+			// can be returned to the pool.
+			origMatrixes[i] = make(Matrix, len(matrixes[i]))
+			copy(origMatrixes[i], matrixes[i])
+		}
+	}
+
+	vectors := make([]Vector, len(exprs)) // Input vectors for the function.
+	// Create an output vector that is as big as the input matrix with
+	// the most time series.
+	biggestLen := 1
+	for i := range exprs {
+		vectors[i] = make(Vector, 0, len(matrixes[i]))
+		if len(matrixes[i]) > biggestLen {
+			biggestLen = len(matrixes[i])
+		}
+	}
+	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
+
+	// The aggregation expects only 1 expression.
+	if len(exprs) != 1 {
+		ev.error(fmt.Errorf("unexpected number of expressions (actual: %d)", len(exprs)))
+	}
+
+	lb := labels.NewBuilder(nil)
+	var buf []byte
+
+	for _, series := range matrixes[0] {
+		metric := series.Metric
+
+		// Compute the grouping key.
+		var groupingKey uint64
+		groupingKey, buf = generateGroupingKey(series.Metric, grouping, without, buf)
+
+		out, ok := seriess[groupingKey]
+
+		// Initialise a new resulting series.
+		if !ok {
+			var groupingMetric labels.Labels
+
+			if without {
+				lb.Reset(metric)
+				lb.Del(grouping...)
+				lb.Del(labels.MetricName)
+				groupingMetric = lb.Labels()
+			} else {
+				groupingMetric = make(labels.Labels, 0, len(grouping))
+				for _, l := range metric {
+					for _, n := range grouping {
+						if l.Name == n {
+							groupingMetric = append(groupingMetric, l)
+							break
+						}
+					}
+				}
+				sort.Sort(groupingMetric)
+			}
+
+			// Initialize output points.
+			// TODO use getPointSlice() and then grow it if cap() is not enough.
+			// points := getPointSlice(numSteps)
+			// points = points[0:numSteps]
+			points := make([]Point, numSteps)
+			idx := 0
+
+			for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+				points[idx].T = ts
+
+				// Initialize to NaN to be able to find out which points have not been
+				// used at all at the end of the series iteration.
+				points[idx].V = math.NaN()
+				idx++
+			}
+
+			out = Series{
+				Metric: groupingMetric,
+				Points: points,
+			}
+			seriess[groupingKey] = out
+		}
+
+		// Sum the series points to the resulting output series.
+		outIdx := 0
+		for _, point := range series.Points {
+			// Advance the output index until we hit the point with the same timestamp.
+			for out.Points[outIdx].T != point.T {
+				outIdx++
+			}
+
+			// Sum aggregation.
+			if math.IsNaN(out.Points[outIdx].V) {
+				out.Points[outIdx].V = point.V
+			} else {
+				out.Points[outIdx].V += point.V
+			}
+		}
+	}
+
+	// Filter out NaN points.
```

Comment:
A point is NaN if there were no series points with that timestamp. We have to filter them out.